PartitioningStrategyFactoryTest.whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance

```
java.lang.AssertionError: expected same:<com.hazelcast.partition.strategy.StringPartitioningStrategy@5a927881> was not:<com.hazelcast.partition.strategy.StringPartitioningStrategy@11bf4347>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotSame(Assert.java:828)
	at org.junit.Assert.assertSame(Assert.java:771)
	at org.junit.Assert.assertSame(Assert.java:782)
	at com.hazelcast.map.impl.PartitioningStrategyFactoryTest.whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance(PartitioningStrategyFactoryTest.java:93)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-ZuluJDK8/com.hazelcast$hazelcast/280/testReport/junit/com.hazelcast.map.impl/PartitioningStrategyFactoryTest/whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance/

**maintenance**
```
java.lang.AssertionError: expected same:<com.hazelcast.partition.strategy.StringPartitioningStrategy@1bf1ce4> was not:<com.hazelcast.partition.strategy.StringPartitioningStrategy@445d49e3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotSame(Assert.java:828)
	at org.junit.Assert.assertSame(Assert.java:771)
	at org.junit.Assert.assertSame(Assert.java:782)
	at com.hazelcast.map.impl.PartitioningStrategyFactoryTest.whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance(PartitioningStrategyFactoryTest.java:93)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-OracleJDK8/com.hazelcast$hazelcast/410/testReport/junit/com.hazelcast.map.impl/PartitioningStrategyFactoryTest/whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance/

```
java.lang.AssertionError: expected same:<com.hazelcast.partition.strategy.StringPartitioningStrategy@54bccadc> was not:<com.hazelcast.partition.strategy.StringPartitioningStrategy@35fdd8f2>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotSame(Assert.java:828)
	at org.junit.Assert.assertSame(Assert.java:771)
	at org.junit.Assert.assertSame(Assert.java:782)
	at com.hazelcast.map.impl.PartitioningStrategyFactoryTest.whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance(PartitioningStrategyFactoryTest.java:93)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-OracleJDK8/com.hazelcast$hazelcast/919/testReport/junit/com.hazelcast.map.impl/PartitioningStrategyFactoryTest/whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance/

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapContainer.java
@@ -178,8 +178,7 @@ public class MapContainer {
     }
 
     private PartitioningStrategy createPartitioningStrategy() {
-        return PartitioningStrategyFactory.getPartitioningStrategy(mapServiceContext.getNodeEngine(),
-                mapConfig.getName(), mapConfig.getPartitioningStrategyConfig());
+        return mapServiceContext.getPartitioningStrategy(mapConfig.getName(), mapConfig.getPartitioningStrategyConfig());
     }
 
     public Indexes getIndexes() {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContext.java
@@ -17,6 +17,7 @@
 package com.hazelcast.map.impl;
 
 import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.PartitioningStrategyConfig;
 import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.map.impl.event.MapEventPublisher;
 import com.hazelcast.map.impl.eviction.ExpirationManager;
@@ -143,5 +144,9 @@ public interface MapServiceContext extends MapServiceContextInterceptorSupport, 
     void incrementOperationStats(long startTime, LocalMapStatsImpl localMapStats, String mapName, Operation operation);
 
     boolean removeMapContainer(MapContainer mapContainer);
+
+    PartitioningStrategy getPartitioningStrategy(String mapName, PartitioningStrategyConfig config);
+
+    void removePartitioningStrategyFromCache(String mapName);
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
@@ -18,6 +18,7 @@ package com.hazelcast.map.impl;
 
 import com.hazelcast.config.Config;
 import com.hazelcast.config.MapConfig;
+import com.hazelcast.config.PartitioningStrategyConfig;
 import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.map.MapInterceptor;
@@ -104,6 +105,7 @@ class MapServiceContextImpl implements MapServiceContext {
     protected final MapQueryEngine mapQueryEngine;
     protected final QueryOptimizer queryOptimizer;
     protected final ContextMutexFactory contextMutexFactory = new ContextMutexFactory();
+    protected final PartitioningStrategyFactory partitioningStrategyFactory;
     protected MapEventPublisher mapEventPublisher;
     protected MapService mapService;
     protected EventService eventService;
@@ -123,6 +125,7 @@ class MapServiceContextImpl implements MapServiceContext {
         this.mapQueryEngine = createMapQueryEngine(queryOptimizer);
         this.eventService = nodeEngine.getEventService();
         this.operationProviders = createOperationProviders();
+        this.partitioningStrategyFactory = new PartitioningStrategyFactory(nodeEngine.getConfigClassLoader());
     }
 
     MapOperationProviders createOperationProviders() {
@@ -597,5 +600,15 @@ class MapServiceContextImpl implements MapServiceContext {
     @Override
     public boolean removeMapContainer(MapContainer mapContainer) {
         return mapContainers.remove(mapContainer.getName(), mapContainer);
+    }
+
+    @Override
+    public PartitioningStrategy getPartitioningStrategy(String mapName, PartitioningStrategyConfig config) {
+        return partitioningStrategyFactory.getPartitioningStrategy(mapName, config);
+    }
+
+    @Override
+    public void removePartitioningStrategyFromCache(String mapName) {
+        partitioningStrategyFactory.removePartitioningStrategyFromCache(mapName);
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/PartitionContainer.java
@@ -164,7 +164,7 @@ public class PartitionContainer {
         if (mapServiceContext.removeMapContainer(mapContainer)) {
             mapContainer.onDestroy();
         }
-        PartitioningStrategyFactory.removePartitioningStrategyFromCache(mapContainer.getName());
+        mapServiceContext.removePartitioningStrategyFromCache(mapContainer.getName());
     }
 
     private void clearLockStore(String name) {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/PartitioningStrategyFactory.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/PartitioningStrategyFactory.java
@@ -19,7 +19,6 @@ package com.hazelcast.map.impl;
 import com.hazelcast.config.PartitioningStrategyConfig;
 import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.nio.ClassLoaderUtil;
-import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.util.ExceptionUtil;
 
 import java.util.concurrent.ConcurrentHashMap;
@@ -29,49 +28,54 @@ import java.util.concurrent.ConcurrentHashMap;
  */
 public final class PartitioningStrategyFactory {
 
-    private static final ConcurrentHashMap<String, PartitioningStrategy> CACHE =
-            new ConcurrentHashMap<String, PartitioningStrategy>();
+    // not private for tests
+    final ConcurrentHashMap<String, PartitioningStrategy> cache = new ConcurrentHashMap<String, PartitioningStrategy>();
 
-    private PartitioningStrategyFactory() {
+    // this is set to the current NodeEngine.getConfigClassLoader
+    private final ClassLoader configClassLoader;
+
+    /**
+     * Construct a new PartitioningStrategyFactory
+     * @param configClassLoader the current {@code NodeEngine}'s {@code configClassLoader}.
+     */
+    public PartitioningStrategyFactory(ClassLoader configClassLoader) {
+        this.configClassLoader = configClassLoader;
     }
 
     /**
-     * Obtain a {@link PartitioningStrategy} for the given {@code NodeEngine} and {@code PartitioningStrategyConfig}. This method
+     * Obtain a {@link PartitioningStrategy} for the given {@code NodeEngine} and {@code mapName}. This method
      * first attempts locating a {@link PartitioningStrategy} in {code config.getPartitioningStrategy()}. If this is {@code null},
      * then looks up its internal cache of partitioning strategies; if one has already been created for the given
      * {@code mapName}, it is returned, otherwise it is instantiated, cached and returned.
-     * @param nodeEngine    Hazelcast NodeEngine
      * @param mapName       Map for which this partitioning strategy is being created
      * @param config        the partitioning strategy configuration
      * @return
      */
-    public static PartitioningStrategy getPartitioningStrategy(NodeEngine nodeEngine, String mapName,
-                                                               PartitioningStrategyConfig config) {
-            PartitioningStrategy strategy = null;
-            if (config != null) {
-                strategy = config.getPartitioningStrategy();
-                if (strategy == null) {
-                    if (CACHE.containsKey(mapName)) {
-                        strategy = CACHE.get(mapName);
-                    } else if (config.getPartitioningStrategyClass() != null) {
-                        try {
-                            strategy = ClassLoaderUtil.newInstance(nodeEngine.getConfigClassLoader(),
-                                    config.getPartitioningStrategyClass());
-                            CACHE.put(mapName, strategy);
-                        } catch (Exception e) {
-                            throw ExceptionUtil.rethrow(e);
-                        }
+    public PartitioningStrategy getPartitioningStrategy(String mapName, PartitioningStrategyConfig config) {
+        PartitioningStrategy strategy = null;
+        if (config != null) {
+            strategy = config.getPartitioningStrategy();
+            if (strategy == null) {
+                if (cache.containsKey(mapName)) {
+                    strategy = cache.get(mapName);
+                } else if (config.getPartitioningStrategyClass() != null) {
+                    try {
+                        strategy = ClassLoaderUtil.newInstance(configClassLoader, config.getPartitioningStrategyClass());
+                        cache.put(mapName, strategy);
+                    } catch (Exception e) {
+                        throw ExceptionUtil.rethrow(e);
                     }
                 }
             }
-            return strategy;
+        }
+        return strategy;
     }
 
     /**
      * Remove the cached {@code PartitioningStrategy} from the internal cache, if it exists.
      * @param mapName name of the map whose partitioning strategy will be removed from internal cache
      */
-    public static void removePartitioningStrategyFromCache(String mapName) {
-        CACHE.remove(mapName);
+    public void removePartitioningStrategyFromCache(String mapName) {
+        cache.remove(mapName);
     }
 }

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java
@@ -42,7 +42,6 @@ import com.hazelcast.map.impl.MapEntries;
 import com.hazelcast.map.impl.MapService;
 import com.hazelcast.map.impl.MapServiceContext;
 import com.hazelcast.map.impl.PartitionContainer;
-import com.hazelcast.map.impl.PartitioningStrategyFactory;
 import com.hazelcast.map.impl.event.MapEventPublisher;
 import com.hazelcast.map.impl.nearcache.NearCacheInvalidator;
 import com.hazelcast.map.impl.nearcache.NearCacheProvider;
@@ -195,8 +194,8 @@ abstract class MapProxySupport extends AbstractDistributedObject<MapService> imp
 
         this.mapServiceContext = service.getMapServiceContext();
         this.mapConfig = mapConfig;
-        this.partitionStrategy = PartitioningStrategyFactory.getPartitioningStrategy(nodeEngine,
-                mapConfig.getName(), mapConfig.getPartitioningStrategyConfig());
+        this.partitionStrategy = mapServiceContext.getPartitioningStrategy(mapConfig.getName(),
+                mapConfig.getPartitioningStrategyConfig());
         this.localMapStats = mapServiceContext.getLocalMapStatsProvider().getLocalMapStatsImpl(name);
         this.partitionService = getNodeEngine().getPartitionService();
         this.lockSupport = new LockProxySupport(new DefaultObjectNamespace(SERVICE_NAME, name),

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/PartitioningStrategyFactoryTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/PartitioningStrategyFactoryTest.java
@@ -16,28 +16,35 @@
 
 package com.hazelcast.map.impl;
 
+import com.hazelcast.config.Config;
+import com.hazelcast.config.MapConfig;
 import com.hazelcast.config.PartitioningStrategyConfig;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
 import com.hazelcast.core.PartitioningStrategy;
+import com.hazelcast.map.impl.proxy.MapProxyImpl;
 import com.hazelcast.partition.strategy.StringPartitioningStrategy;
-import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
-import org.hamcrest.core.IsSame;
+import com.hazelcast.util.RootCauseMatcher;
+import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.rules.ExpectedException;
+import org.junit.rules.TestName;
 import org.junit.runner.RunWith;
-import org.mockito.Mockito;
 
-import static com.hazelcast.map.impl.PartitioningStrategyFactory.getPartitioningStrategy;
+import java.util.UUID;
+
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertSame;
-import static org.mockito.Mockito.mock;
+import static org.junit.Assert.assertTrue;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
@@ -46,36 +53,43 @@ public class PartitioningStrategyFactoryTest extends HazelcastTestSupport {
     @Rule
     public ExpectedException expectedException = ExpectedException.none();
 
-    @Test
-    public void testUtilityClassWithPrivateConstructor() {
-        assertUtilityConstructor(PartitioningStrategyFactory.class);
+    @Rule
+    public TestName testName = new TestName();
+
+    private IMap<String, String>  map;
+    private String mapName;
+    private PartitioningStrategyFactory partitioningStrategyFactory;
+
+    @Before
+    public void setup() {
+        mapName = testName.getMethodName();
+        HazelcastInstance hazelcastInstance = createHazelcastInstance();
+        map = hazelcastInstance.getMap(mapName);
+        MapProxyImpl mapProxy = (MapProxyImpl) map;
+        MapService mapService = (MapService) mapProxy.getService();
+        MapServiceContextImpl mapServiceContext = (MapServiceContextImpl) mapService.getMapServiceContext();
+        partitioningStrategyFactory = mapServiceContext.partitioningStrategyFactory;
     }
 
     @Test
     public void whenConfigNull_getPartitioningStrategy_returnsNull() {
-        HazelcastInstance hz = createHazelcastInstance();
-        NodeEngine nodeEngine = getNodeEngineImpl(hz);
-        PartitioningStrategy partitioningStrategy = getPartitioningStrategy(nodeEngine, "map", null);
+        PartitioningStrategy partitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, null);
         assertNull(partitioningStrategy);
     }
 
     @Test
     public void whenPartitioningStrategyDefined_getPartitioningStrategy_returnsSameInstance() {
-        HazelcastInstance hz = createHazelcastInstance();
-        NodeEngine nodeEngine = getNodeEngineImpl(hz);
         PartitioningStrategy configuredPartitioningStrategy = new StringPartitioningStrategy();
         PartitioningStrategyConfig cfg = new PartitioningStrategyConfig(configuredPartitioningStrategy);
-        PartitioningStrategy partitioningStrategy = getPartitioningStrategy(nodeEngine, "map", cfg);
+        PartitioningStrategy partitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, cfg);
         assertSame(configuredPartitioningStrategy, partitioningStrategy);
     }
 
     @Test
     public void whenPartitioningStrategyClassDefined_getPartitioningStrategy_returnsNewInstance() {
-        HazelcastInstance hz = createHazelcastInstance();
-        NodeEngine nodeEngine = getNodeEngineImpl(hz);
         PartitioningStrategyConfig cfg = new PartitioningStrategyConfig();
         cfg.setPartitioningStrategyClass("com.hazelcast.partition.strategy.StringPartitioningStrategy");
-        PartitioningStrategy partitioningStrategy = getPartitioningStrategy(nodeEngine, "map", cfg);
+        PartitioningStrategy partitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, cfg);
         assertEquals(StringPartitioningStrategy.class, partitioningStrategy.getClass());
     }
 
@@ -83,13 +97,12 @@ public class PartitioningStrategyFactoryTest extends HazelcastTestSupport {
     // strategy for the same map name retrieves the same instance
     @Test
     public void whenStrategyForMapAlreadyDefined_getPartitioningStrategy_returnsSameInstance() {
-        HazelcastInstance hz = createHazelcastInstance();
-        NodeEngine nodeEngine = getNodeEngineImpl(hz);
         PartitioningStrategyConfig cfg = new PartitioningStrategyConfig();
         cfg.setPartitioningStrategyClass("com.hazelcast.partition.strategy.StringPartitioningStrategy");
         // when we have already obtained the partitioning strategy for a given map name
-        PartitioningStrategy partitioningStrategy = getPartitioningStrategy(nodeEngine, "map", cfg);
-        PartitioningStrategy cachedPartitioningStrategy = getPartitioningStrategy(nodeEngine, "map", cfg);
+        PartitioningStrategy partitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, cfg);
+        // then once we get it again with the same arguments, we retrieve the same instance
+        PartitioningStrategy cachedPartitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, cfg);
         assertSame(partitioningStrategy, cachedPartitioningStrategy);
     }
 
@@ -97,17 +110,50 @@ public class PartitioningStrategyFactoryTest extends HazelcastTestSupport {
     // then the exception is rethrown (the same if it is a RuntimeException, otherwise it is peeled,
     // see ExceptionUtil.rethrow for all the details).
     @Test
-    public void whenStrategyInstantiationThrowsException_getSamePartitioningStrategy_rethrowsException() {
-        NodeEngine nodeEngine = mock(NodeEngine.class);
-        RuntimeException e = new RuntimeException("expected exception");
-        Mockito.when(nodeEngine.getConfigClassLoader()).thenThrow(e);
+    public void whenStrategyInstantiationThrowsException_getPartitioningStrategy_rethrowsException() {
+        PartitioningStrategyConfig cfg = new PartitioningStrategyConfig();
+        cfg.setPartitioningStrategyClass("NonExistentPartitioningStrategy");
 
+        // while attempting to get partitioning strategy, ClassNotFound exception will be thrown and wrapped in HazelcastException
+        expectedException.expect(new RootCauseMatcher(ClassNotFoundException.class));
+        // use a random UUID as map name, to avoid obtaining the PartitioningStrategy from cache.
+        partitioningStrategyFactory.getPartitioningStrategy(UUID.randomUUID().toString(), cfg);
+    }
+
+    @Test
+    public void whenRemoveInvoked_entryIsRemovedFromCache() {
         PartitioningStrategyConfig cfg = new PartitioningStrategyConfig();
         cfg.setPartitioningStrategyClass("com.hazelcast.partition.strategy.StringPartitioningStrategy");
+        PartitioningStrategy partitioningStrategy = partitioningStrategyFactory.getPartitioningStrategy(mapName, cfg);
+        partitioningStrategyFactory.removePartitioningStrategyFromCache(mapName);
+        assertFalse(partitioningStrategyFactory.cache.containsKey(mapName));
+    }
 
-        // while attempting to get partitioning strategy, exception will be thrown
-        expectedException.expect(new IsSame<RuntimeException>(e));
-        getPartitioningStrategy(nodeEngine,
-                "whenStrategyInstantiationThrowsException_getSamePartitioningStrategy_rethrowsException", cfg);
+    @Test
+    public void whenMapDestroyed_entryIsRemovedFromCache() {
+        map.put("a", "b");
+        // at this point, the PartitioningStrategyFactory#cache should have an entry for this map config
+        assertTrue("Key should exist in cache", partitioningStrategyFactory.cache.containsKey(mapName));
+
+        map.destroy();
+
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
+                assertFalse("Key should have been removed from cache", partitioningStrategyFactory.cache.containsKey(mapName));
+            }
+        }, 20);
+    }
+
+    @Override
+    protected Config getConfig() {
+        Config config = new Config();
+        MapConfig mapConfig = new MapConfig(mapName);
+        PartitioningStrategyConfig partitioningStrategyConfig = new PartitioningStrategyConfig();
+        partitioningStrategyConfig.setPartitioningStrategyClass("com.hazelcast.partition.strategy.StringPartitioningStrategy");
+        mapConfig.setPartitioningStrategyConfig(partitioningStrategyConfig);
+        config.addMapConfig(mapConfig);
+        return config;
     }
 }

